clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java

```diff
@@ -1111,6 +1108,86 @@ public void testSubscriptionChangesWithAutoCommitDisabled() {
         consumer.close();
     }
 
+    @Test
+    public void testUnsubscribeWillTriggerPartitionsRevokeWithNoGeneration() {
+        Time time = new MockTime();
+        SubscriptionState subscription = new SubscriptionState(new LogContext(), OffsetResetStrategy.EARLIEST);
+        ConsumerMetadata metadata = createMetadata(subscription);
+        MockClient client = new MockClient(time, metadata);
+
+        initMetadata(client, Collections.singletonMap(topic, 1));
+        Node node = metadata.fetch().nodes().get(0);
+
+        ConsumerPartitionAssignor assignor = new RangeAssignor();
+        KafkaConsumer<String, String> consumer = newConsumer(time, client, subscription, metadata, assignor, false, groupInstanceId);
+
+        initializeSubscription(consumer, getExceptionConsumerRebalanceListener());
+
+        prepareRebalance(client, node, assignor, singletonList(tp0), null);
+
+        try {
+            consumer.updateAssignmentMetadataIfNeeded(time.timer(Long.MAX_VALUE));
+            consumer.poll(Duration.ZERO);
+            fail("Should throw exception");
+        } catch (Throwable e) {
+            assertEquals("Hit partition assign", e.getCause().getMessage());
+        }
+        try {
+            consumer.unsubscribe();
+            fail("Should throw exception");
+        } catch (Throwable e) {
+            assertEquals("Hit partition revoke", e.getCause().getMessage());
+        }
+    }
+
+    @Test
+    public void testUnsubscribeWillTriggerPartitionsLostWithNoGeneration() throws InterruptedException {
+        Time time = new MockTime();
+        SubscriptionState subscription = new SubscriptionState(new LogContext(), OffsetResetStrategy.EARLIEST);
+        ConsumerMetadata metadata = createMetadata(subscription);
+        MockClient client = new MockClient(time, metadata);
+
+        initMetadata(client, Collections.singletonMap(topic, 1));
+        Node node = metadata.fetch().nodes().get(0);
+
+        ConsumerPartitionAssignor assignor = new RangeAssignor();
+        KafkaConsumer<String, String> consumer = newConsumer(time, client, subscription, metadata, assignor, false, groupInstanceId);
+
+        initializeSubscription(consumer, getExceptionConsumerRebalanceListener());
+        Node coordinator = prepareRebalance(client, node, assignor, singletonList(tp0), null);
+
+        try {
+            consumer.updateAssignmentMetadataIfNeeded(time.timer(Long.MAX_VALUE));
+            consumer.poll(Duration.ZERO);
+            fail("Should throw exception");
+        } catch (Throwable e) {
+            assertEquals("Hit partition assign", e.getCause().getMessage());
+        }
+
+        AtomicBoolean heartbeatReceived = prepareHeartbeatResponse(client, coordinator, Errors.UNKNOWN_MEMBER_ID);
+
+        time.sleep(heartbeatIntervalMs);
+        Thread.sleep(heartbeatIntervalMs);
```

Comment:
Instead of waiting on thread to sleep, we can just wait for `heartbeatReceived.get()` in `waitForCondition`. Since we are using a mock time here we do not need to actually sleep that long, the hb thread should see the result immediately in the next call.